### PR TITLE
Support django 2.0

### PR DESCRIPTION
**Pull request by Surgo**
- Support django 2.0 and add tests
- Remove EOL versions (1.8, 1.9, 1.10 - [roadmap](https://www.djangoproject.com/weblog/2015/jun/25/roadmap/))
- Run only supported python versions for each django versions ([doc](https://docs.djangoproject.com/en/2.0/faq/install/#what-python-version-can-i-use-with-django))
- Use tox for testing (flake8, readme tests are run only once)
- Cache wheels in travis cache mechanism

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,8 +1,108 @@
...
+.tox/
+.coverage
+.coverage.*
 .cache
-.tox
+nosetests.xml
+coverage.xml
+*.cover
+.hypothesis/
+.pytest_cache/
```
Comment (Surgo) [inline .gitignore]: `.pytest_cache/` directory did not ignored. Generate from github's ignore boilerplate.

File: .travis.yml [hunk 2]
```diff
@@ -1,21 +1,38 @@
...
-  - 3.5
-  - 3.6
-env:
-  - DJANGO_VERSION=1.8.*
-  - DJANGO_VERSION=1.9.*
-  - DJANGO_VERSION=1.10.*
-  - DJANGO_VERSION=1.11.*
+dist: trusty
+language: python
+cache: pip
```
Comment (Surgo) [inline .travis.yml]: Cache pip and wheels

File: .travis.yml [hunk 3]
```diff
@@ -1,21 +1,38 @@
...
+  - env: TOXENV=py35-djmaster
+    python: 3.5
+  - env: TOXENV=py36-djmaster
+    python: 3.6
+  - env: TOXENV=flake8
+  - env: TOXENV=isort
+  - env: TOXENV=readme
+  - env: TOXENV=check-manifest
+  allow_failures:
+  - env: TOXENV=py34-dj111
```
Comment (Surgo) [inline .travis.yml]: In README.rst, py34 does not supported.

File: django_redshift_backend/base.py [hunk 4]
```diff
@@ -165,9 +164,10 @@ def create_model(self, model):
                 definition += " %s" % col_type_suffix
             params.extend(extra_params)
             # FK
-            if field.rel and field.db_constraint:
-                to_table = field.rel.to._meta.db_table
-                to_column = field.rel.to._meta.get_field(field.rel.field_name).column
+            print(dir(field))
+            if field.remote_field and field.db_constraint:
```
Comment (Surgo) [inline django_redshift_backend/base.py]: `Field.rel` is removed. Use `remote_field`.

https://docs.djangoproject.com/en/2.0/releases/1.9/#field-rel-changes

File: tests/testapp/models.py [hunk 5]
```diff
@@ -23,5 +23,5 @@ class TestParentModel(models.Model):
 
 
 class TestChildModel(models.Model):
-    parent = models.ForeignKey(TestParentModel)
+    parent = models.ForeignKey(TestParentModel, on_delete=models.CASCADE)
```
Comment (Surgo) [inline tests/testapp/models.py]: `on_delete` is always required.

https://docs.djangoproject.com/en/2.0/releases/1.9/#foreignkey-and-onetoonefield-on-delete-argument

File: django_redshift_backend/base.py (hunk 4 above)
Comment (Surgo) [reply django_redshift_backend/base.py]: Oh... print func

File: django_redshift_backend/base.py [hunk 6]
```diff
@@ -165,9 +164,10 @@ def create_model(self, model):
                 definition += " %s" % col_type_suffix
             params.extend(extra_params)
             # FK
-            if field.rel and field.db_constraint:
-                to_table = field.rel.to._meta.db_table
-                to_column = field.rel.to._meta.get_field(field.rel.field_name).column
+            if field.remote_field and field.db_constraint:
```
Comment (Surgo) [inline django_redshift_backend/base.py]: `Field.rel` is deprecated in 1.9 and removed in 2.0.

https://docs.djangoproject.com/en/2.0/releases/1.9/#field-rel-changes

**Comment by Surgo** [comment]
Would you like to setup pypi deployment via travis?

https://docs.travis-ci.com/user/deployment/pypi/

**Comment by shimizukawa** [comment]
Thanks!
I'll take a look after #30.

**Comment by shimizukawa** [comment]
> Would you like to setup pypi deployment via travis?

Good to know, thanks!

If I joined to jazzband on #30, the community mechanism will provides auto-releasing feature via Travis https://jazzband.co/about/releases. I guess the feature is same as you suggested https://docs.travis-ci.com/user/deployment/pypi/.

File: .gitignore [hunk 7]
```diff
@@ -1,8 +1,108 @@
+### https://raw.github.com/github/gitignore/be1d0e0a4cd9734c5c9541413eb1d777fbb74c6e/Python.gitignore
```
Comment (shimizukawa) [inline .gitignore]: I don't think this long ignore list is needed for this project.

File: .travis.yml [hunk 8]
```diff
@@ -1,21 +1,38 @@
...
+dist: trusty
+language: python
+cache: pip
+matrix:
+  fast_finish: true
+  include:
+  - env: TOXENV=py27-dj111
+    python: 2.7
+  - env: TOXENV=py34-dj111
+    python: 3.4
```
Comment (shimizukawa) [inline .travis.yml]: Why do you add python-3.4 testing?
django-redshift-backend doesn't support python-3.4.

File: tox.ini [hunk 9]
```diff
@@ -1,23 +1,43 @@
 [tox]
-envlist = flake8,py{27,35,36}-django{18,19,110,111}
+envlist =
+    py{27,34,35,36}-dj111,
```
Comment (shimizukawa) [inline tox.ini]: (again) Why py34? django-redshift-backend doesn't support python-3.4

File: tox.ini [hunk 10]
```diff
@@ -1,23 +1,43 @@
...
+    check-manifest
 skipsdist = True
 
 [testenv]
 deps =
-    django18: Django>=1.8,<1.9
-    django19: Django>=1.9,<1.10
-    django110: Django>=1.10,<1.11
-    django111: Django>=1.11,<1.12
-    -rtest-requires.txt
```
Comment (shimizukawa) [inline tox.ini]: If you don't think this line is not required, `test-requires.txt` file is not needed anymore.
And please explain your purpose.

File: MANIFEST.in [hunk 11]
```diff
@@ -0,0 +1,17 @@
...
+
+global-exclude __pycache__
+global-exclude *.py[co]
+global-exclude *.cfg
+global-exclude *.ini
+global-exclude *.txt
+global-exclude *.yaml
+
+prune .github
+prune tests
```
Comment (shimizukawa) [inline MANIFEST.in]: I've never seen `check-manifest` tool before, but if this content is the result of using the `check-manifest` tool, I think it's too much.

File: README.rst [hunk 12]
```diff
@@ -133,6 +129,7 @@ Bug Fixes:
 * #15,#21: More compat with redshift: not use CHECK. Thanks to Vasil Vangelovski.
 * #18: Fix error on migration with django-1.9 or later that raises AttributeError
   of 'sql_create_table_unique'.
+* #27: annotate() does not work on Django-1.9 and later. Thanks to Takayuki Hirai.
```
Comment (shimizukawa) [inline README.rst]: Oops, Thanks!

File: django_redshift_backend/base.py (hunk 6 above)
Comment (shimizukawa) [reply django_redshift_backend/base.py]: 👍

File: tox.ini [hunk 13]
```diff
@@ -1,23 +1,43 @@
...
-    flake8 django_redshift_backend
-    python setup.py check -s -r
+basepython = python3
+deps = flake8
+commands = flake8
+
+[testenv:isort]
+basepython = python3
+deps = isort
+commands = isort --recursive --verbose --check-only --diff django_redshift_backend tests setup.py
```
Comment (shimizukawa) [inline tox.ini]: Great! I didn't know the isort tool!

File: tox.ini [hunk 14]
```diff
@@ -1,23 +1,43 @@
...
+
+[testenv:readme]
+basepython = python3
+deps = readme_renderer
```
Comment (shimizukawa) [inline tox.ini]: 👍

File: tests/testapp/models.py (hunk 5 above)
Comment (shimizukawa) [reply tests/testapp/models.py]: 👍

File: setup.py [hunk 15]
```diff
@@ -32,9 +32,6 @@ def read(filename):
         'Programming Language :: Python :: 3.5',
         'Programming Language :: Python :: 3.6',
         'Framework :: Django',
-        'Framework :: Django :: 1.8',
-        'Framework :: Django :: 1.9',
-        'Framework :: Django :: 1.10',
         'Framework :: Django :: 1.11',
```
Comment (shimizukawa) [inline setup.py]: We must have a Django 2.0 line.

File: setup.py [hunk 16]
```diff
@@ -1,7 +1,7 @@
-from setuptools import setup, find_packages
+from setuptools import find_packages, setup
 
 requires = [
-    'django',
+    'Django>=1.11',
```
Comment (shimizukawa) [inline setup.py]: +0. I think some users can use a part of this backend library if this version limitation is not exist (if they lucky). However I don’t oppose version constraints.

File: tox.ini (hunk 10 above)
Comment (Surgo) [reply tox.ini]: 1. Depend libs are different for each python versions. 

    Like
    ```
    docutils<0.13  # 0.13 + py27 cause 'setup.py check' error. see: https://bugs.python.org/issue23063
    mock
    ```

2. Depends libs are different for each test (like flake8, isort, readme etc). This effective for test performance.

I made PR for test performance after merge this PR.

File: setup.py [hunk 17]
```diff
@@ -1,43 +1,40 @@
-from setuptools import setup, find_packages
```
Comment (Surgo) [inline setup.py]: This file use crlf?

**Comment by Surgo** [comment]
@shimizukawa 

> Can you split this PR into separated PRs for each purpose (I suggest as following)? I would like to merge from what we have resolved by making necessary discussions about each PR.

Sorry, I split PR.

This PR is for drop EOL version and support 2.0.

File: test-requires.txt [hunk 18]
```diff
@@ -1,7 +1,6 @@
 -f./wheelhouse
 psycopg2==2.5.5
```
Comment (Surgo) [inline test-requires.txt]: Caught psycopg/psycopg2#594 error on python 3.6

Comment (Surgo) [reply test-requires.txt]: Maybe version up required.

File: setup.py (hunk 17 above)
Comment (Surgo) [reply setup.py]: at 4cf14520c42845936e33fde7af1b004a7b1159c8

File: test-requires.txt [hunk 19]
```diff
@@ -1,7 +1,6 @@
...
 pytest
-mock
```
Comment (Surgo) [inline test-requires.txt]: Its standard lib

File: .travis.yml [hunk 20]
```diff
@@ -5,12 +5,15 @@ python:
...
   - DJANGO_VERSION=1.10.*
   - DJANGO_VERSION=1.11.*
+  - DJANGO_VERSION=2.0.*
+matrix:
+  exclude:
+  - python: 2.7
+    env: DJANGO_VERSION=2.0.*
 install:
-  - pip install -U pip wheel
+  - pip install -U pip==9.0.* wheel
```
Comment (Surgo) [inline .travis.yml]: https://travis-ci.org/shimizukawa/django-redshift-backend/builds/367612528

Cannot install psycopg2 using python3.6 and pip 10.0.0 from wheel.

Its try to download from pypi and build from source. and its failed

```
2.11s$ pip install -U -r test-requires.txt
Looking in links: /home/travis/build/shimizukawa/django-redshift-backend/./wheelhouse
Obtaining file:///home/travis/build/shimizukawa/django-redshift-backend (from -r test-requires.txt (line 6))
Collecting psycopg2==2.5.5 (from -r test-requires.txt (line 2))
  Downloading https://files.pythonhosted.org/packages/9b/60/f4c79e73a69ded145880bcf4f98eeed741af12c62c5ddc89b754602b1807/psycopg2-2.5.5.tar.gz (779kB)
    100% |████████████████████████████████| 788kB 17.2MB/s 
    Complete output from command python setup.py egg_info:
    running egg_info
    creating pip-egg-info/psycopg2.egg-info
    writing pip-egg-info/psycopg2.egg-info/PKG-INFO
    writing dependency_links to pip-egg-info/psycopg2.egg-info/dependency_links.txt
    writing top-level names to pip-egg-info/psycopg2.egg-info/top_level.txt
    writing manifest file 'pip-egg-info/psycopg2.egg-info/SOURCES.txt'
    Error: could not determine PostgreSQL version from '10.1'
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-install-ykdfkudz/psycopg2/
```

Comment (Surgo) [reply .travis.yml]: Is master passes build?

**Comment by Surgo** [comment]
Close it and made small PRs.

**Pull request closed without merging**